src/uuid.ts

```diff
@@ -1,12 +1,39 @@
+import { intRangeWithEngine } from "./int-range.js";
 import { Engine, defaultEngine } from "./utils.js"
 
 const uuid = (engine: Engine = defaultEngine): string => {
   // Credit @Alexey Silin from https://gist.github.com/1308368
-  let a: any = ''
-  let b: any = ''
-  // eslint-disable-next-line
-  for (b = a = ''; a++ < 36; b += ~a % 5 | a * 3 & 4 ? (a ^ 15 ? 8 ^ engine() * (a ^ 20 ? 16 : 4) : 4).toString(16) : '-') { }
-  return b
+	// Creates a random v4 UUID of the form xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx,
+	// where each x is a random hexadecimal digit from 0 to f, and y is a random
+	// hexadecimal digit from 8 to b.
+	const randomRange = intRangeWithEngine(engine)
+	const defaultGenerator = (): string => randomRange(0, 15).toString(16)
+	const groups: object[] = [
+		{ length: 8 },
+		{ length: 4 },
+		{
+			length: 4,
+			generator: (index: number) => index === 0 ? 4 : defaultGenerator()
+		},
+		{
+			length: 4,
+			generator: (index: number) =>
+				(index === 0 ? randomRange(8, 11).toString(16) : defaultGenerator())
+		},
+		{ length: 12 }
+	];
+	return groups.map(
+		({ length, generator }: any) => {
+			const group: string[] = [];
+			for (let index = 0; index < length; index++) {
+				group.push(generator !== undefined
+					? generator(index)
+					: defaultGenerator()
+				)
+			}
+			return group.join('')
```

Comment:
this can all be just 1 reduce statement.

```typescript
return groups.reduce((acc, { length, generator }) => {
  const groupStr = ''
  for (let index = 0; index < length; index++) {
    const val = generator ? generator(index) : defaultGenerator()
    groupStr += val
  }
  acc += `-${groupStr}`
  return acc
}, '')
```